### ISSUE TITLE
Ref parameter incorrectly decorated with `noalias` attribute

### ISSUE THREAD

**Issue by RalfJung** [rust-lang/rust#63787]
The following library:
```rust
use std::cell::*;

pub fn break_it(rc: &RefCell<i32>, r: Ref<'_, i32>) {
    // `r` has a shared reference, it is passed in as argument and hence
    // a barrier is added that marks this memory as read-only for the entire
    // duration of this function.
    drop(r);
    // *oops* here we can mutate that memory.
    *rc.borrow_mut() = 2;
}
```
[generates IR](https://play.rust-lang.org/?version=nightly&mode=release&edition=2018&gist=459fd0c36c90d82ff9abde235febfc7b) for `break_it` that starts like
```llvm
; playground::break_it
; Function Attrs: noinline nonlazybind uwtable
define void @_ZN10playground8break_it17hb8a9beeb2affda19E(
  { i64, i32 }* align 8 dereferenceable(16) %rc,
  i32* noalias nocapture readonly align 4 dereferenceable(4) %r.0,
  i64* nocapture align 8 dereferenceable(8) %r.1)
unnamed_addr #1 personality i32 (i32, i32, i64, %"unwind::libunwind::_Unwind_Exception"*, %"unwind::libunwind::_Unwind_Context"*)* @rust_eh_personality {
```
Note the `noalias`. That is incorrect, this function violates the noalias assumptions by mutating the data `r` points to through a pointer (`rc`) not derived from `c`. Here's a caller causing bad aliasing:
```rust
pub fn main() {
    let rc = RefCell::new(0);
    break_it(&rc, rc.borrow())
}
```

`RefMut` has the same problem when `-Zmutable-noalias` is set.

Cc @rkruppe @comex

**Comment by gnzlbg** [comment]
cc @eddyb

**Comment by bjorn3** [comment]
Maybe pour a `UnsafeCell` into `Ref`?

**Comment by RalfJung** [comment]
It should probably use a raw pointer (something I have been arguing for for years now ;). For `RefMut`, `UnsafeCell` wouldn't even help.

**Comment by gnzlbg** [comment]
Changing `Ref` value field to `value: &'b std::cell::UnsafeCell<T>` fixes this.

**Comment by RalfJung** [comment]
Also see https://github.com/rust-lang/rust/issues/60076: `vec_deque::Drain` has a similar problem.

> Changing Ref value field to value: &'b std::cell::UnsafeCell<T> fixes this.

It does, but I think a raw pointer is the better fix.

**Comment by gnzlbg** [comment]
> It does, but I think a raw pointer is the better fix.

Why? 

The value field inside `RefCell` is of type `UnsafeCell<T>`, not `T`, so IMO `Ref` has a bug by using `&T` instead of `&UnsafeCell<T>`. Using a raw pointer would mean that we loose the dereferenceable annotation.

**Comment by comex** [comment]
Proof of concept miscompilation: [playground link](https://play.rust-lang.org/?version=stable&mode=release&edition=2018&gist=55c92c46b47f071175f0b5e902793773)

**Comment by eddyb** [comment]
In https://github.com/rust-lang/unsafe-code-guidelines/issues/125#issuecomment-523876722 I asked:
> Is `Ref` having `noalias readonly` on the data reference a problem because, unlike a regular reference, it can be invalidated by `Drop`?
>
> Could we then rely on the presence of a `Drop` `impl` to disable `noalias readonly`, or should `Ref` itself use a raw pointer or `&UnsafeCell`?

But it seems like the preferred solution is to change `Ref` to not contain a reference.

> For `RefMut`, `UnsafeCell` wouldn't even help.

@RalfJung Why not? It would have to be `&UnsafeCell`, not `&mut UnsafeCell`, but I expect it to work.

**Comment by RalfJung** [comment]
> It would have to be &UnsafeCell

Hm okay that might work. That seems real ugly though.

> Why?

IMO raw pointers are more honest, as laid out [years ago](https://github.com/rust-lang/rust-memory-model/issues/5). The lifetime in that reference is a lie.

**Comment by RalfJung** [comment]
Nominated for lang team as well.

@rust-lang/lang the interesting question from a lang team perspective here is to decide whether this is a library bug or a language bug. I.e., is the code wrong and the `noalias` right, or vice versa? Do we want non-`repr(transparent)`-structs containing references to be marked `noalias`?

**Comment by comex** [comment]
> Could we then rely on the presence of a `Drop` `impl` to disable `noalias readonly`, or should `Ref` itself use a raw pointer or `&UnsafeCell`?

Well, we have other types such as `Box`, `Vec`,  etc. that have `Drop` impls but still want `noalias`.

With such types, the pointer *can* be deallocated by `Drop` and later reused, but... for one thing, the deallocation function takes the pointer itself as an argument, which is treated as the pointer escaping.  If the function makes a new allocation and get the same pointer value back, the new pointer can be said to be 'based on' the pointer it had deallocated – depending on the allocator implementation, it often actually is, and when it isn't, LLVM generally can't prove that.

More importantly, `__rust_alloc` has its return value declared `noalias`, so LLVM will always assume that newly allocated pointers don't alias other pointers, regardless of whether the other pointers are themselves `noalias`.  This attribute can in fact lead to miscompilations: it's one of the issues that the 'twin allocation' paper tried to address.  There might be a case for removing it.  But as long as it's there, removing `noalias` from `Box` and co. wouldn't make a difference, regardless of the 'based on' question.

**Comment by rkruppe** [comment]
> With such types, the pointer _can_ be deallocated by `Drop` and later reused, but... for one thing, the deallocation function takes the pointer itself as an argument, which is treated as the pointer escaping. If the function makes a new allocation and get the same pointer value back, the new pointer can be said to be 'based on' the pointer it had deallocated – depending on the allocator implementation, it often actually is, and when it isn't, LLVM generally can't prove that.

I would argue for the opposite: if an allocation is freed and the underlying memory is recycled to satisfy another allocation later on, the later allocation should result in a pointer of different provenance from the one that was freed, regardless of the physical address reuse. Each new allocation should be considered a fresh allocation for the purposes of the abstract machine (to justify the different provenance). This is somewhat realized today by allocation functions returning `noalias` pointers, as you noted yourself.

This should not lead to any issues because all uses of the "old" allocation must happen before it is freed, and all uses of the new allocation must happen after it was allocated. If these two time spans don't overlap, everything is find. It's a bit weird that pointers to the same memory address but with different provenance are floating around next to each other, but this can happen even in safe code with e.g. int-to-pointer casts or `wrapping_offset`. And if these two time spans do overlap (i.e., the new allocation is used before the old one is freed), then the new allocation is made before the old one is freed, so it can't reuse the same address.

Note that all of this is completely independent of allocator internals and data structures, which are neither relevant nor usually visible to optimizers.

Freeing an allocation to which one has a safe pointer or reference is problematic for other reasons, namely `dereferenceable` (see e.g. https://internals.rust-lang.org/t/is-it-possible-to-be-memory-safe-with-deallocated-self/8457/12), which we add not only to references but also to `Box` and other owned pointers. We do need to solve this somehow, but I don't see a similar issue for `noalias`.

> More importantly, `__rust_alloc` has its return value declared `noalias`, so LLVM will always assume that newly allocated pointers don't alias other pointers, regardless of whether the other pointers are themselves `noalias`. This attribute can in fact lead to miscompilations: it's one of the issues that the 'twin allocation' paper tried to address. There might be a case for removing it. But as long as it's there, removing `noalias` from `Box` and co. wouldn't make a difference, regardless of the 'based on' question.

I wrote above why I think `noalias` on those functions is fine. I also don't recall seeing (and couldn't find with a quick scan just now) any mention of such an issue in the twin allocation paper. Leaving aside that `noalias` literally doesn't occur at all (IIRC it's out of scope / left for future work), what *is* there goes in quite the opposite direction: if my reading of figure 5 is correct, `malloc` always returns a *fresh, logical* pointer!

> * Though Vec can't actually get `noalias` treatment, at least on x86, because the ABI passes it by a pointer. You know more about that than I do.

FWIW this is more a consequence of deliberate choices in the Rust ABI(s) to pass structs of two scalars as those scalars but not do the same for structs with more fields. It's neither platform-specific nor otherwise externally-imposed. Also, if we had usable "local `noalias`" (e.g., via metadata on loads) then this wouldn't matter anyway.

**Comment by comex** [comment]
> Note that all of this is completely independent of allocator internals and data structures, which are neither relevant nor usually visible to optimizers.

They're visible if LTO is turned on, but unless the allocator is trivial, the optimizer is nowhere near smart enough to notice that newly allocated pointers might not be 'based on' the previously freed versions.  However, assuming that the optimizer will never be smart enough to notice something is a risky strategy in the long term. :)

> I wrote above why I think `noalias` on those functions is fine. I also don't recall seeing (and couldn't find with a quick scan just now) any mention of such an issue in the twin allocation paper. Leaving aside that `noalias` literally doesn't occur at all (IIRC it's out of scope / left for future work), what _is_ there goes in quite the opposite direction: if my reading of figure 5 is correct, `malloc` always returns a _fresh, logical_ pointer!

Well, it's a combination of two issues mentioned in the paper.

In 4.6 it discusses the problems with propagating pointer equalities in GVN – that is, in a code path where `a == b` must have previously evaluated to true, replacing `a` with `b`.  This is the same optimization I mentioned in my proof-of-concept for the `Ref` issue earlier in this thread.

In 4.8, it discusses how newly allocated pointers can be equal to previously freed ones, albeit mostly in the context of wanting to optimize the comparison to `false`.  However, one of the approaches it describes would be an alternative way to avoid this issue:

> Another solution is to define comparisons with freed blocks as UB (as C and C++ standards do)

To be more concrete, here is an example 'miscompilation' (but not really, because it cheats) based on propagating pointer equalities across a deallocation and reallocation.  Under the model proposed by the twin allocation paper, the miscompilation is the optimizer's fault rather than Rust's; I believe the authors' proof-of-concept LLVM modification would have prevented it from occurring, by disabling GVN for pointers outside of limited scenarios where it's known to be safe.
[Playground link](https://play.rust-lang.org/?version=stable&mode=release&edition=2018&gist=d12e4c7ae23663ead103fb39501aa238)

The steps:
- Start with a `Box<i32>` parameter, `x`.  Store `2` to it.
- Save `x`'s pointer value as a raw pointer variable.
- Deallocate `x`.
- Obtain a new `Box<i32>`, `y`, from another function; it has value `4`.
- Assert that `y`'s pointer value equals the previously saved one from `x`.
- Load from `y`.

Since LLVM knows the new and old boxes' pointers are equal at the point of the load, it treats it as a load from the old box... and since the old box  is `noalias`, it assumes that nobody else could have written to it after it stored a `2`, so it optimizes the load to always return `2`.

As I said, the example cheats.  This optimization can only happen if the pointer doesn't escape from the function, and calling `__rust_dealloc` on it counts as it escaping.  So instead of having the function free the `Box` directly, it uses a callback which reconstitutes the `Box` from a previously saved pointer and frees it – which is UB.

However, there are other circumstances where this could *theoretically* happen without UB.  Suppose that instead of a `Box`, the function takes a reference-counted pointer to immutable data, and that pointer is hypothetically `noalias`.  `Rc` won't work for this purpose: the pointer it contains cannot be `noalias` because it's used to access the reference count, so it's not fully immutable.  But you could imagine a type where the reference count is stored separately from the data itself:
```rust
struct DetachedRc<T> {
    refcount: *mut i32,
    ptr: *const T,
}
```
In this case, `ptr` could hypothetically be legally marked as `noalias` (with some suitable wrapper type that doesn't currently exist), assuming that `T` does not contain `UnsafeCell`.

But, assuming `DetachedRc`'s drop routine is inlined into the caller, it does not necessarily cause `ptr` to escape.  As long as there's an additional reference, dropping it will only mutate the `refcount`, and then some other code can drop the last reference and actually free it.  You could use this to construct an example similar to my cheating one but without the UB, yet still vulnerable to the unsound optimization.

(Some practical issues: (1) you would have to make LLVM somehow know statically that the victim function cannot drop the last reference, perhaps by asserting on the refcount, and (2) you'd be working with immutable data, whereas my PoC involves mutation; `assert!(*x == 2);` *ought* to work just as well as `*x = 2;` to let LLVM know that future dereferences should return 2, but it didn't seem to work when I tested it.)

**Comment by RalfJung** [comment]
Another argument for using raw pointers: not only `noalias` but also `derefencable` is wrong. Consider this example:
```rust
use std::cell::*;

pub fn break_it(rc: &RefCell<Option<Box<i32>>>, r: Ref<'_, i32>) {
    drop(r);
    *rc.borrow_mut() = None;
}

pub fn main() {
    let rc = RefCell::new(Some(Box::new(0)));
    break_it(&rc, Ref::map(rc.borrow(), |x| &**x.as_ref().unwrap()))
}
```
The memory `r` points to is deallocated while `break_it` still runs.

Of course, this could also be viewed as more evidence that `UnsafeCell` should inhibit `dereferencable` as well.

---

> Could we then rely on the presence of a Drop impl to disable noalias readonly, or should Ref itself use a raw pointer or &UnsafeCell?

I don't think that would be sufficient in general. For example, consider a silly version of `Ref` that does *not* implement `drop`, but provides a manual "destroy" method. That would still be a sound data structure, but adding `noalias` / `dereferencable` would be just as wrong as with the real `Ref`.

One could say this is silly, but I am not sure interpreting "fuzzy" signals like "is there a `Drop` impl" for properties like this is a good idea. It also doesn't solve the `dereferencable` problem.

Btw, `readonly` is fine I think. That just means that no writes happen through this particular pointer, right?

@comex 

> More importantly, __rust_alloc has its return value declared noalias, so LLVM will always assume that newly allocated pointers don't alias other pointers, regardless of whether the other pointers are themselves noalias.

That's a very different `noalias` though -- its behavior in return position is not really the same as its behavior in argument position. Notably, the return position one has "infinite scope", whereas the argument position one only has an effect during the one function call.

Custom allocators are a very underexplored topic, and the LLVM "twin allocation" paper did not explore them either, that one assumed a built-in allocator. The paper does not even mention `noalias`. It is all about modelling allocators. LLVM uses `noalias` to indicate that a function behaves "allocator-like", but that doesn't mean that the "twin allocation" model is the right one for custom allocators, it might well only apply to the built-in one.

> the deallocation function takes the pointer itself as an argument, which is treated as the pointer escaping

"escaping" is not an Abstract Machine operation, it is an approximation used by compiler. So it can never be used to argue for correctness.

@rkruppe 

> FWIW this is more a consequence of deliberate choices in the Rust ABI(s) to pass structs of two scalars as those scalars but not do the same for structs with more fields.

Does this mean what I wrote at https://github.com/rust-lang/rust/issues/60076#issuecomment-523490943 is wrong? I saw a `noalias` for a `vec_deque::Drain` and concluded that it is also affected by this bug.

---

I have no idea why you are talking about custom allocators and return-position `noalias`, how is that relevant for the discussion?

**Comment by rkruppe** [comment]
> Another argument for using raw pointers: not only `noalias` but also `derefencable` is wrong. Consider this example:

👍 

> Does this mean what I wrote at [#60076 (comment)](https://github.com/rust-lang/rust/issues/60076#issuecomment-523490943) is wrong? I saw a `noalias` for a `vec_deque::Drain` and concluded that it is also affected by this bug.

Oh, yeah, the `noalias` that shows up in the IR when passing `vec_deque::Drain` by value is *on a pointer to the `Drain`* (which points to a stack slot allocated for passing the parameter by value, which indeed isn't aliased), not on an actual data pointer contained in the iterator.

> I have no idea why you are talking about custom allocators and return-position noalias, how is that relevant for the discussion?

I don't know why @comex brought it up here, I don't think it's related (nor am I convinced there is any issue to begin with). In any case I've opened https://github.com/rust-lang/unsafe-code-guidelines/issues/196 to discuss it separately.

**Comment by gnzlbg** [comment]
> @rust-lang/lang the interesting question from a lang team perspective here is to decide whether this is a library bug or a language bug. I.e., is the code wrong and the noalias right, or vice versa? 

Can we answer this question without an aliasing model ?

Independently of who is at fault, fixing this at the "library level", by using raw pointers, would be a small and localized change, that we could justify by trying to avoid relying on unspecified details of the aliasing model.

Fixing this at the language level, by removing `noalias` for all infringing types, would require changing the ABI of non-repr-transparent Adts containing fields of reference type. I can imagine that we could always pass these by memory, or that we could remove `noalias` annotations of reference fields when passing these by value. But these changes would make `struct Foo<'a, T>(&'a mut T);` subtly different from `&'a mut T`, and could result in users destructuring Adts manually before function calls to regain `noalias`.

**Comment by RalfJung** [comment]
> Can we answer this question without an aliasing model ?

This would be putting a constraint on the aliasing model. We already have many those constraints (e.g. from code where everyone agrees `noalias` is correct). This would be adding one more.

> would require changing the ABI of non-repr-transparent Adts containing fields of reference type.

Not really. At least, I don't view whether `noalias` is added as part of the ABI. This is a callee choice. There is no representation change.

**Comment by gnzlbg** [comment]
> E.g., we say that Option<&mut T> in Rust and *T in C have the same ABI -- but one gets noalias and the other does not.

They have the same calling convention (the values are passed in the same way from caller to callee and vice-versa), but they are different types, just like `u32` and `NonZeroU32`, and this means that one can't use them interchangeably. The C type that one can probably use interchangeably with `Option<&mut T>` is `T* restrict`, and this matters because `T*` and `T* restrict` and not [compatible C types](https://en.cppreference.com/w/c/language/type#Compatible_types). See [this libc soundness bug](https://github.com/rust-lang/libc/pull/1346#issuecomment-495537142), and the [tracking ctest issue](https://github.com/gnzlbg/ctest/issues/68).

> This is a callee choice. 

This choice must be made depending on Layout/ABI/PointerKind/etc.

**Comment by comex** [comment]
> I don't know why @comex brought it up here, I don't think it's related (nor am I convinced there is any issue to begin with). In any case I've opened rust-lang/unsafe-code-guidelines#196 to discuss it separately.

Sorry, I got pretty far into the weeds.  My point is:
- `noalias` is wrong for `Ref` arguments, because you can drop a `Ref`, then get a unique reference to the same object from another source.
- Is `noalias` also wrong for `Box` arguments, because you can drop a `Box`, then get a unique reference to a new allocation which happens to be at the same address?
- Answer: No, because LLVM would assume old allocations don’t alias new even if `Box` arguments weren’t `noalias`, thanks to the allocator’s return value being `noalias`.  Therefore we may as well keep `Box` arguments noalias.
- That is, unless we decide to reconsider the allocator return value being `noalias`, which I guess we can discuss in the other thread.

**Comment by comex** [comment]
> Both of your examples seem to crucially rely on this GVN bug. Please correct me if I'm wrong (I've not fully digested either yet), but if that's true, I don't understand what either of them have to do with allocation returning `noalias` pointers. It seems just yet another variation of pointing out that pointers to equal addresses can be non-interchangeable for actual memory accesses, just with a lot of unnecessary extra complications.

Pretty much.

I just wouldn't characterize it as "yet another variation".  If anything, it's one of two variations.  The twin allocation paper creates pointers with the same address but different provenance by using two allocations that are adjacent in space (but exist at the same time).  My second example does so using two allocations that are separated in time (but at the same address in space).

My first example with (`Ref`) uses two pointers that derive from the same allocation and thus have the same provenance.  It does rely on GVN to produce an actual miscompilation, but GVN in this particular case would be sound accordin to the conditions in the twin-allocation paper, and other optimizations besides GVN could theoretically cause problems too.  I think we're all in agreement that `noalias` is incorrect for `Ref`.

But going back to the other example – I tried to make it clear that the problem in that example is *likely* LLVM's fault, not Rust's.  `noalias` is *likely* fine on `Box` and on the allocator's return value.  However, I wanted to leave the door open for other conclusions, because I have no evidence that LLVM will actually implement the GVN limitations proposed in the twin allocation paper.

For one thing, they don't seem to be in much of a hurry to implement any sort of formal model.

For another, the model from the twin allocation paper is only one possibility among others.  Even if I assumed that LLVM would end up making its optimizations sound with respect to some formal model, that wouldn't automatically justify basing Rust's `noalias` decisions on the twin allocation model in particular.

My chain of reasoning in the other thread was a bit muddled, but the part about "escaping" is essentially a sketch of an alternative solution for *temporal* aliasing, compared to how the twin allocation model solves it.  Namely, the allocator is modeled such that when it reuses an address, the pointer it returns is 'based on' the argument previously passed to `free`.  The benefit of this model is that it makes unrestricted GVN on pointers sound with respect to temporal aliasing (only).  Therefore, it would be potentially useful:

    - As part of a formal model, in combination with some other solution to spatial aliasing that also allows unrestricted or less-restricted GVN compared to the twin allocation model.

      I don't know what that solution would look like, exactly, and it would probably have other drawbacks.  But the question of what model to use is up to LLVM anyway.  I'm just exploring possibilities.

      But also:

    - As a conservative approach for the time being, in order to mitigate some of the miscompilations caused by LLVM's current unrestricted GVN.

      I don't know how useful that would actually be.  It wouldn't fix all miscompilations, so perhaps it would just be a pointless band-aid, especially because GVN-based miscompilations tend to only occur in contrived examples anyway.

But for the record, under that model, `Box`-typed arguments could still be `noalias`, but the allocator's return value could not be, nor could arguments of certain other smart pointer types such as my hypothetical `DetachedRc`.

Finally:

> All that said, it is likely true that when you stop treating alloc/dealloc functions as built-ins all the time and start to (selectively) look at the C or Rust code that implements them through the lens of the language's general memory model, you'll run into trouble. But that's not at all surprising to me (those functions are deliberately treated as magic after all), and it's not related to the examples you've given, or to `noalias` annotation and address reuse in particular. For example, if you inline only the `free` part of a malloc-free pair and the allocator implementation stores metadata inline, then the code from `free` will appear to perform OOB accesses when it pokes the allocation metadata.

LLVM can, in fact, inline an alloc/dealloc function under the right circumstances.  So the potential for "trouble" exists, mitigated by LLVM's lack of sophistication.  If you say this is off topic – sure.  I originally only mentioned it in passing.

**Comment by comex** [comment]
[Note: before this reply, I posted and then deleted a reply that was meant for [another thread](https://github.com/rust-lang/unsafe-code-guidelines/issues/196).]

> They have the same calling convention (the values are passed in the same way from caller to callee and vice-versa), but they are different types, just like `u32` and `NonZeroU32`, and this means that one can't use them interchangeably. The C type that one can probably use interchangeably with `Option<&mut T>` is `T* restrict`, and this matters because `T*` and `T* restrict` and not [compatible C types](https://en.cppreference.com/w/c/language/type#Compatible_types). See [this libc soundness bug](https://github.com/rust-lang/libc/pull/1346#issuecomment-495537142), and the [tracking ctest issue](https://github.com/gnzlbg/ctest/issues/68).

Is there any ABI that actually represents `T *` and `T *restrict` differently?  I suppose it could come up that ABIs that check the type signature out of band in some way, like asm.js using a different array for indirect calls depending on the function type.  But neither GCC nor Clang complains if you convert a function pointer of type `void (*a)(void *restrict)` to one of type `void (*a)(void *)`, even with `-Wall -Wextra -pedantic`, and they're represented the same in LLVM IR (so it doesn't affect asm.js as compiled by LLVM).

**Comment by gnzlbg** [comment]
>  But neither GCC nor Clang complains if you convert a function pointer of type void (*a)(void *restrict) to one of type void (*a)(void *), even with -Wall -Wextra -pedantic, and they're represented the same in LLVM IR (so it doesn't affect asm.js as compiled by LLVM).

I'm not sure how good their checking is, but they do complain here, and in both directions: https://godbolt.org/z/88SKj2

**Comment by RalfJung** [comment]
@comex
> Is noalias also wrong for Box arguments, because you can drop a Box, then get a unique reference to a new allocation which happens to be at the same address?

Oh I see. Yes that is a good, if only remotely related, question.  Answering it in detail is non-trivial without a proper spec for `noalias` in argument position. But `noalias` does not imply `dereferencable`, and it does also not mean that the *pointer* doesn't alias -- it means that *accesses done through the pointer* don't alias (conflict) with accesses done through other pointers, so they can be reordered.

The new allocation, even if it is at the same address, really has different pointers pointing to it (recorded through their provenance), so those will not be confused.

@gnzlbg 
I think we wall agree that it is legal to have C FFI with type `Option<&mut T>` where C uses a pointer, right? That means C pointers and `Option<&mut T>` are ABI-compatible as far as their value and memory ABI is concerned. This is even true with `&mut T` and C pointers. So surely the same is true on the Rust side between `&mut T` and a newtype around `&mut T` that drops `noalias`.

Whatever C specifies for `restrict` is not very relevant for us, unless LLVM explicitly does the same for `noalias`.

**Comment by gnzlbg** [comment]
Consider this soundness issue https://github.com/rust-lang/rust/issues/46188 applied to Rust ([godbolt](https://llvm.godbolt.org/z/JvQt7n)):

```rust
pub mod bad {
    extern "Rust" { fn foo(x: Option<&mut i32>); }
}

extern "Rust" { fn foo(y: *mut i32); }
pub unsafe fn baz(x: *mut i32) { foo(x) }
```

LLVM will "optimize" the declaration of `crate::foo` making its `y: *mut i32` noalias..., because there is a different declaration in `bad::foo` to the same item that has all those attributes due to its type being `Option<&mut i32>`.

If the type of `crate::foo` were `fn(Option<&mut T>)`, this wouldn't be an issue.

Is this optimization sound ? I don't know. If we translate this example to C, this program has UB. 

> I think we wall agree that it is legal to have C FFI with type Option<&mut T> where C uses a pointer, right?

Now imagine what happens when you move the declaration of `fn crate::foo` to a C translation unit that you link with the `bad` module and LTO kicks in. In LLVM-IR it doesn't matter that one module comes from Rust and the other from C, and because we have used `Option<&mut T>` in Rust to interface with C code expecting `T*`, the declaration in C suddenly gets turned into an "`Option<&mut T>`" as well. I haven't managed to make LLVM propagate these attributes to definitions, but this might just be pure luck. 

---

So is it legal to have C FFI with type Option<&mut T> where C uses a pointer? 

I don't know. _I wish_ it was so that we can use Rust types to "enrich" C function declarations with information that C APIs express in comments. 

We can definitely say that Rust's `&mut T`, `&T`, `Option<&T>`, `*mut T`, ... (for `T: Sized`) have all the same size, align, fields, and call ABI, as C's `T*`, `T* const volatile restrict`, ... They do not have the same niches (e.g. `nonnull`) and other attributes like `dereferenceable`, `noalias`, etc. and this results in miscompilations today.

This all goes back to "What is an ABI?", e.g., are niches part of the ABI ? So when some document currently says that `Option<&mut T>` is ABI compatible with C pointers, that isn't necessarily wrong. It depends on what that document means by "ABI". What's wrong is to say that one can substitute one for the other. Without LTO, that might be true as well, since there is no way for LLVM to tell and miss optimize your program, but with LTO, what matters is what optimizations are sound for LLVM-IR and in this case, it is unclear whether these optimizations are correct or not.

I suppose that worst case we could stop emitting `noalias`, `dereferenceable`, etc. in function declarations, which might or might not have an impact on optimizations. But we could also say, just like C does, that the behavior of providing a declaration that does not match a definition is undefined. E.g. if you use `Option<&mut T>` in Rust, and `T*` in C, that C code using `T*` might get optimized as if it were using `Option<&mut T>`. If there are C comments saying that this is ok, then you are fine. But if that C code is called with aliasing pointers somewhere else, then.. boom.

**Comment by RalfJung** [comment]
> Now imagine what happens when you move the declaration of fn crate::foo to a C translation unit that you link with the bad module and LTO kicks in. The only thing that matters there is whether this optimization is sound in LLVM-IR, and today, because we have used Option<&mut T> in Rust to interface with C code expecting T*, the declaration in C suddenly gets turned into an "Option<&mut T>" as well.

In other words, when declaring an FFI function with references, you have to make sure that the body of that function actually respects the noalias guarantees expressed by those references. That seems like a fair requirement to make -- but this is a very subtle interaction indeed. :(

> if C code calls such a declaration with aliasing pointers, then the code might have been miscompiled.

Nit: `noalias` is not about whether the pointers alias but about whether the function uses the pointers in an aliasing way.

Anyway, all of that seems *far* off-topic for the topic here. If you think these issues are important, I suggest opening a new issue somewhere (not sure where... UCG or reference or nomicon, I guess).

In Rust, modulo issues like https://github.com/rust-lang/rust/issues/28179, every function only has one declaration, so I don't think we have to worry about these multi-declaration issues. Transmutes back and forth and even casting function pointers should still be fine -- or do function pointers also have attributes like `noalias`?

**Comment by gnzlbg** [comment]
> In Rust, modulo issues like #28179, every function only has one declaration

Yeah, without `#[no_mangle]`, if you were linking two Rust libraries and wanted to use a function from one in the other by declaring it, you would need to `extern "Rust" { fn $mangled_name(...) -> ... ; } `. That could be made to work correctly and reliably, e.g., by generating `$mangled_name` using a `build.rs` (IIRC our mangling scheme was RFC'ed? not sure how stable this is), but feels like a long shot. 

>  or do function pointers also have attributes like noalias?

Not in LLVM-IR so AFAICT this cannot happen either (when calling a function pointer, we use the Rust types to manually annotate the call, but that's it).

**Comment by joshtriplett** [comment]
We discussed that further in today's lang team meeting.

We don't see any obvious downsides to using raw pointers for `Ref`.

@withoutboats made the point that we may want to use `NonNull`, and may wish to comment on that further.

That aside, we do think that changing Ref seems like the best option here.

**Comment by Centril** [comment]
(@RalfJung may want to clone the current implementation as a Miri test to use as a canary or something.)

**Comment by bstrie** [comment]
Is this issue still applicable? When trying the original example on 1.56-nightly (which AFAIK has mutable-noalias on by default) I get the following:
```
; playground::break_it
; Function Attrs: nonlazybind uwtable
define internal void @_ZN10playground8break_it17h8910f8f2e3383d74E({ i64, i32 }* align 8 dereferenceable(16) %rc, i32* align 4 dereferenceable(4) %r.0, i64* align 8 dereferenceable(8) %r.1) unnamed_addr #1 personality i32 (i32, i32, i64, %"unwind::libunwind::_Unwind_Exception"*, %"unwind::libunwind::_Unwind_Context"*)* @rust_eh_personality !dbg !797 {
st
```

**Comment by eddyb** [comment]
AFAIK `noalias readonly` (i.e. not mutable, not even by the rest of the world) is what the example in the issue description is about (also I don't think we ever omitted it for `&T` references), and that seems to be gone (at least from your LLVM IR snippet).
The issue description mentions `RefMut` too, but you'd have to make your own example by replacing `Ref` with `RefMut` (but it does indeed look like even in that case, and even with `-Zmutable-noalias`, the `noalias` is gone: https://godbolt.org/z/M1KK4a6Ks).

(I've also edited both the issue description and your comment to add syntax highlighting via \`\`\`llvm - sadly I really wish LLVM did any kind of multi-line formatting, it's really annoying to have to use horizontal scroll to read these things)

**Comment by RalfJung** [comment]
Interesting! So when did this change, and what exactly are the rules for `noalias` emission these days? I'm not aware of any lang team decision here, so this might be mostly an accident.

**Comment by bjorn3** [comment]
I believe the last changes to these rules was in https://github.com/rust-lang/rust/pull/82834: https://github.com/rust-lang/rust/blob/cfc856acf3112d241bd4de55ec91df5aef66c352/compiler/rustc_middle/src/ty/layout.rs#L2340-L2368

**Comment by RalfJung** [comment]
Ah... `noalias` is now only added in optimized builds. And with the `-O` flag, everything seems to behave as before.

**Comment by bstrie** [comment]
By "behave as before", you mean you're still seeing an invalid `noalias`? If I try your original code example with 1.56-nightly in release mode, and add in a `#[inline(never)]` so that the function doesn't boil away, I get this, without any `noalias`:
```llvm
; playground::break_it
; Function Attrs: noinline nonlazybind uwtable
define internal fastcc void @_ZN10playground8break_it17hc2b3a560f5f197e6E({ i64, i32 }* align 8 dereferenceable(16) %rc, i64* nocapture align 8 dereferenceable(8) %r.1) unnamed_addr #0 personality i32 (i32, i32, i64, %"unwind::libunwind::_Unwind_Exception"*, %"unwind::libunwind::_Unwind_Context"*)* @rust_eh_personality {
```
What is the exact output that you're seeing, and with which version?

**Comment by eddyb** [comment]
Make sure to add `-C no-prepopulate-passes`, otherwise you get the fully optimized inference, including e.g. `readnone` on unused `fn` parameters. I haven't actually checked it, tho, just a pitfall to be aware of.

**Comment by RalfJung** [comment]
@bstrie I [tried this](https://godbolt.org/z/cd7dafrfh) and the 2nd argument of `break_it_ref` becomes
```
i32* noalias readonly align 4 dereferenceable(4) %0
```
Though strangely there is a third argument...? Ah I guess this is using `ScalarPair` layout, passing the two fields of `Ref` as separate arguments. The first one is the relevant one (`value: &'b T`), assuming no reordering happened. (The second one points to an `UnsafeCell` so it cannot have `noalias` here anyway.)

Same for `break_it_ref_mut`:
```
i32* noalias align 4 dereferenceable(4) %0
```

> Make sure to add -C no-prepopulate-passes, otherwise you get the fully optimized inference, including e.g. readnone on unused fn parameters. I haven't actually checked it, tho, just a pitfall to be aware of.

I would be very surprised if LLVM ever added `noalias` automatically. I don't think this would be sound, certainly not in this case.

**Comment by RalfJung** [comment]
Indeed the original example for some reason looks very different, I think LLVM might be optimizing away an unused function argument? [Here's a playground reproducer though](https://play.rust-lang.org/?version=nightly&mode=release&edition=2018&gist=459fd0c36c90d82ff9abde235febfc7b). I will update the original text.

**Comment by eddyb** [comment]
> I believe the last changes to these rules was in #82834:
> 
> <https://github.com/rust-lang/rust/blob/cfc856acf3112d241bd4de55ec91df5aef66c352/compiler/rustc_middle/src/ty/layout.rs#L2340-L2368>

@bjorn3 Ah just opened this on desktop and I can see the code - I'm not really happy with non-codegen parts of the compiler relying on opt-level, but I can kinda see how you couldn't avoid the `is_freeze` query.
Maybe we should speed up the query instead of trying to bypass it? I wonder what the common case is. (cc @nikic)

**Comment by eddyb** [comment]
@RalfJung Ah I remember now, I removed the body because for some reason I was getting things I didn't want to see in the LLVM IR, codegen'd, and ended up with something like this: https://godbolt.org/z/aooY9jbMa
You can see `nocapture readnone` in there, since it's an unused `&T`.

If you add `-C no-prepopulate-passes` however, you do get the expected attributes: https://godbolt.org/z/sznos8aGz.
And `-C no-prepopulate-passes` should also help with the binary crate version, since you'll be able to see what we gave to LLVM, before it got to touch it at all.

> I would be very surprised if LLVM ever added `noalias` automatically.

Ah, fair, I was mostly noticing the `readnone`.

**Comment by bstrie** [comment]
So to be clear, is the fix for this as straightforward as changing `Ref`'s `value` field from `&T` to `*const T`, and `RefMut`'s from `&mut T` to `*mut T`?

**Comment by RalfJung** [comment]
Yes I think that would be the best fix (and I've [argued for that change](https://github.com/rust-lang/rust-memory-model/blob/master/litmus_tests/refcell_ref.md) even before we knew about this problem^^)

**Comment by BurntSushi** [comment]
cc @rust-lang/libs What's stopping us from fixing this? It looks like a low hanging unsound bug. :)

**Comment by RalfJung** [comment]
One problem is deciding whether this should be fixed on the lang level or libs level...

**Comment by BurntSushi** [comment]
It looks like we could at least fix this at the libs level without much difficulty, right? Then the lang team can decide whether they want to fix this on their end or not. If they do, I assume we could then revert the library fix.

Although I will say that I don't quite grok the lang side fix for this.

**Comment by RalfJung** [comment]
Yeah, I think there are several good reasons to use raw pointers (or rather, `NonNull`) rather than references here.

On the lang side, we could not add `noalias` to newtyped references... but that might also be a bad idea since it makes newtypes less powerful.

**Comment by cuviper** [comment]
Please see #97027.

**Comment by cuviper** [comment]
Now that the library fix has landed, are we still considering lang changes, or can we close this?

**Comment by RalfJung** [comment]
I think the lang side of this is not final yet, but this is not the right issue to track that. We have https://github.com/rust-lang/unsafe-code-guidelines/issues/125 and this will have to be part of the discussion of the aliasing model, once we get there.
